[Francis Tan] iP

Francis-Tan opened the pull request:
# Duke
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's

- Text-based
- Easy to learn
- ~~Fast~~ _SUPER FAST_ to use

All you need to do is

1. Download it from [here](https://github.com/Francis-Tan/ip/releases)
2. Save it into a folder
3. Add your tasks
4. Let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x] Organize tasks into 3 types: Todos, Deadlines, Events
- [x] Mark and unmark any task as done
- [x] Type in words and find matching tasks
- [x] Specify date and optionally time for Deadlines and Events
- [ ] Reminders (coming soon)

If you're a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

Francis-Tan committed:
Level-1

Francis-Tan committed:
Level-2

Francis-Tan committed:
Added message when listing tasks

Francis-Tan committed:
Level-3

Francis-Tan committed:
Level-4

Francis-Tan committed:
Leading and trailing whitespaces in input are ignored. Commands, instead of reading the input itself, now read words from the input split by " ", to make code more readable

Francis-Tan committed:
Set up automated text UI testing through I/O redirection

Francis-Tan committed:
Exceptions and their messages implemented for missing descriptions/flags/datetimes, invalid args for un/mark

Francis-Tan committed:
Implemented delete operation with invalid task number exceptions

Francis-Tan committed:
Refactored duke by factoring out common functionality among the commands, along with more javadocs

Francis-Tan committed:
[A-Packages] Put all classes in a duke folder containing folders for tasks and services

Francis-Tan committed:
[A-MoreOOP] Divided Duke's functionalities into different classes, shifted static methods over to them

Francis-Tan committed:
[A-Gradle] Gradle set up

Francis-Tan committed:
[A-JUnit] Implemented some unit tests for Todo and Parser

Francis-Tan committed:
configured build.gradle to create a duke jar file

Francis-Tan committed:
Implemented saving and loading in storage

Francis-Tan committed:
Changed timing field to only accept the format d/M/yyyy followed by an optional (h:mm)am/pm, and the description shows the month in english

Francis-Tan committed:
Merge branch 'branch-Level-7'

Francis-Tan committed:
Merge branch 'branch-Level-8'

Francis-Tan committed:
Updated deadline and event timing getters to fix saving and loading under new timing format

Francis-Tan committed:
Added a few javadocs

Francis-Tan committed:
Changed a name to comply with coding standard

Francis-Tan committed:
Added find command, which lists tasks whose description contains the given keyword

Francis-Tan committed:
Merge branch 'branch-A-JavaDoc'

Francis-Tan committed:
Resolved conflict over variable name

Francis-Tan committed:
Merge branch 'branch-Level-9'

Francis-Tan committed:
Implement checkstyle and fix violations of coding standard

rui-han-crh commented on src/main/java/duke/services/UI.java:
```diff
@@ -0,0 +1,34 @@
+package duke.services;
+
+/** Handles display */
+public class UI {
```
Perhaps the name of this class would be better written as Ui, to conform to the coding standard?

rui-han-crh commented on src/main/java/duke/services/Parser.java:
```diff
@@ -0,0 +1,173 @@
+        while (!(words.length == 1 && words[0].equals("bye"))) {
+            if (words.length > 0) {
+                try {
+                    if (words.length == 1 && words[0].equals("list")) {
+                        TaskList.listTasks(); //could put words.length == 1 cases all here
+                    } else if (words.length == 1 && words[0].equals("SAVE")) {
+                        Storage.wipeDataOnExit(false);
+                    } else if (words.length == 1 && words[0].equals("WIPE")) {
+                        Storage.wipeDataOnExit(true);
+                    } else if (words[0].equals("todo")) {
+                        TaskList.addTodo(words);
+                    } else if (words[0].equals("deadline")) {
+                        TaskList.addDeadline(words);
+                    } else if (words[0].equals("event")) {
+                        TaskList.addEvent(words);
+                    } else if (words[0].equals("mark")) {
+                        TaskList.markTaskAsDone(words);
+                    } else if (words[0].equals("unmark")) {
+                        TaskList.markTaskAsNotDone(words);
+                    } else if (words[0].equals("delete")) {
+                        TaskList.deleteTask(words);
+                    } else if (words[0].equals("find")) {
+                        TaskList.findTasksContainingKeyword(words);
+                    } else {
+                        UI.sayLines(new String[]{"I'm sorry, I don't know what that means"});
+                    }
```
Perhaps a switch statement would be more conventional for readibility here?

rui-han-crh commented on src/main/java/duke/services/Parser.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    public static void handleUserInputs() {
+        Scanner inputScanner = new Scanner(System.in);
+        String[] words = Arrays.stream(inputScanner.nextLine().strip().split(" ")).toArray(String[]::new);
```
Nice application of declarative input consumption and splitting here! :)

rui-han-crh commented on src/test/java/duke/services/ParserTest.java:
```diff
@@ -0,0 +1,56 @@
+        assertEquals(Parser.getDescription(deadline1, "/by"), "finish   essay");
+        assertEquals(Parser.getTiming(deadline1, "/by"), "tmr night    or ask 4 extension");
+        //does it throw exception for missing all other arguments?
+        assertThrows(IllegalArgumentException.class, () ->
```
I like how you also decided to test the exceptions here!

But this test case may be testing too many things at the same time, perhaps split these tests into more subgroups? Like
`testGetDescription_event`, `testGetDescription_deadline`, `testGetDescriptionAndDate_throwsIllegalArgumentException`, etc

rui-han-crh reviewed (commented):
Very nice! Code is clean and readable, good amount of test cases to cover many bases.

Francis-Tan committed:
Implement GUI

Francis-Tan committed:
Improve code quality

Refactored code by making Duke, rather than the parser, decide how to
respond to inputs, and merged UI with Duke, since UI was in charge of
too little functionality that Duke could also handle reasonably.
Broke down a very long test case into smaller test cases.

Francis-Tan committed:
Add a few small asserts and modifications

Francis-Tan committed:
Merge pull request #3 from Francis-Tan/branch-A-CodeQuality

Improve code quality

Francis-Tan committed:
Merge branch 'branch-A-Assertions'

# Conflicts:
#	src/main/java/duke/services/Duke.java
#	src/main/java/duke/services/Parser.java
#	src/main/java/duke/services/Storage.java

Francis-Tan committed:
Refactor out Ui class from Duke's reply functionality for clarity

Francis-Tan committed:
Implement sort command to sort tasks in ascending order of time
and number of time attributes

Francis-Tan committed:
Merge branch 'branch-C-Sort'

Francis-Tan committed:
Update readme and add Ui.png

Francis-Tan committed:
Update version to v0.2
